### PR TITLE
[Serve] Remove `VersionedReplica` abstract class

### PR DESCRIPTION
**Pull request by shrekris-anyscale**
## Why are these changes needed?

This change removes the `VersionedReplica` abstract class. The only class that inherits from `VersionedReplica` is `DeploymentReplica`, so there's no need to keep it.

This change also replaces all `VersionedReplica` type hints with `DeploymentReplica`.

## Related issue number

N/A

## Checks

- [X] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [X] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - This change relies on existing tests.

### CODE REVIEW

File: python/ray/serve/tests/unit/test_deployment_state.py [hunk 1]
```diff
@@ -332,22 +331,27 @@ def create_deployment_state_manager(
         dead_replicas_context.clear()
 
 
-def replica(version: Optional[DeploymentVersion] = None) -> VersionedReplica:
-    if version is None:
-        version = DeploymentVersion(get_random_string(), DeploymentConfig(), {})
+class FakeDeploymentReplica(DeploymentReplica):
```
Comment (GeneDer) [inline python/ray/serve/tests/unit/test_deployment_state.py:334]: Nitpick: As a fake implementation we probably don't want it to inherit from the real `DeploymentReplica`. We should instead implement any method with the fake implementation if needed.

File: python/ray/serve/tests/unit/test_deployment_state.py [hunk 2]
```diff
@@ -332,22 +331,27 @@ def create_deployment_state_manager(
...
+        return self._version
 
-        def update_state(self, state):
-            pass
+    def update_state(self, state):
+        pass
+
+
+def replica(version: Optional[DeploymentVersion] = None) -> FakeDeploymentReplica:
+    if version is None:
```
Comment (GeneDer) [inline python/ray/serve/tests/unit/test_deployment_state.py:351]: Non-blocker: Can write it as 
```
version = version or DeploymentVersion(get_random_string(), DeploymentConfig(), {})
```
😛

**Review by GeneDer — COMMENTED** [review]
Was there a historical reason we have `VersionedReplica` in the first place? I'm fine deleting it for now, but was also wondering if there will be a future we add it back for other types of replicas 🤔

File: python/ray/serve/tests/unit/test_deployment_state.py (hunk 1 above)
Comment (shrekris-anyscale) [reply python/ray/serve/tests/unit/test_deployment_state.py:334]: We use the `VersionedReplica` type hint throughout the code, but outside of testing, the object is always a `DeploymentReplica`. It makes the type hints in the code harder to follow, and it makes `jump to definition` less useful.

Comment (shrekris-anyscale) [reply python/ray/serve/tests/unit/test_deployment_state.py:334]: @edoakes @GeneDer what do you think?

Comment (edoakes) [reply python/ray/serve/tests/unit/test_deployment_state.py:334]: don't have that strong of an opinion. I prefer what we have now but if it's a practical inconvenience to you we can merge this

**Review by edoakes — APPROVED** [review]

File: python/ray/serve/tests/unit/test_deployment_state.py (hunk 1 above)
Comment (GeneDer) [reply python/ray/serve/tests/unit/test_deployment_state.py:334]: +1 don't have strong opinion on taking `VersionedReplica` out. But maybe we should at least have `FakeDeploymentReplica` not inherit from `DeploymentReplica`

Comment (shrekris-anyscale) [reply python/ray/serve/tests/unit/test_deployment_state.py]: Sounds good, I've removed the inheritance, and I'll merge this after it passes tests.

File: python/ray/serve/tests/unit/test_deployment_state.py (hunk 2 above)
Comment (shrekris-anyscale) [reply python/ray/serve/tests/unit/test_deployment_state.py]: I did so.

**Review by GeneDer — APPROVED** [review]
LGTM! Thanks for addressing my comments Shreyas!

**Pull request merged**